src/fluence/jgf/jgf.go

```diff
@@ -85,260 +106,125 @@ func (g *FluxJGF) MakeEdge(source string, target string, contains string) {
 		},
 	}
 	g.Graph.Edges = append(g.Graph.Edges, newedge)
-	if contains == ContainsRelation {
-		tnode := g.NodeMap[target]
-		tnode.Metadata.Paths[containmentKey] = g.NodeMap[source].Metadata.Paths[containmentKey] + "/" + tnode.Metadata.Name
-	}
 }
 
 // MakeSubnet creates a subnet for the graph
-func (g *FluxJGF) MakeSubnet(index int64, ip string) string {
-	newnode := node{
-		Id: fmt.Sprintf("%d", g.Elements),
+// The name is typically the ip address
+func (g *FluxJGF) MakeSubnet(name string) Node {
+
+	// Get a resource counter for the subnet
+	resource := g.Resources.getCounter(name, SubnetType)
+	subpath := resource.NameWithIndex()
+	return g.makeNewNode(resource, subpath, defaultUnit, defaultSize)
+}
+
+// makeNewNode is a shared function to make a new node from a resource spec
+// subpath is the subpath to add to the graph root, e.g., <cluster>/<subpath>
+// Since there is some variability to this structure, it is assembled by
+// the calling function
+func (g *FluxJGF) makeNewNode(
```

Comment:
This is good progress, but I think we'll need a more general way to create "nodes" in case the cluster has types we don't include here, or flow-based resources like power system status.

Reply:
We could just call this function directly for general types - the function itself is generic (and then the wrappers, for example, `MakeSubnet` help to get the resource and set defaults.

I do have a more generic way - the [jsongraph-go](https://github.com/converged-computing/jsongraph-go) library implements generic JGF but still would need a handler for counters, etc. We would still likely need to have some kind of wrapper to create different types. That's [what I use](https://github.com/converged-computing/jsongraph-go/network/dependents) for compspec-go, rainbow, and all of our flex projects because it makes it really easy to instantiate a graph.

Either approach can handle custom or generic types, and we'd just need a slightly tweaked function. I was conservative about making changes here as to not introduce too many in one PR - the goal of this PR was to reduce the redundancy and I'm definitely good to keep working on this further (and use another strategy entirely)! Let's chat about it next time we have a fluence discussion.